aten/src/ATen/cuda/CUDAGraph.cpp

```diff
@@ -65,6 +67,16 @@ CUDAGraph::CUDAGraph()
 
 void CUDAGraph::capture_begin(MempoolId_t pool/*=0*/) {
 #if !defined(USE_ROCM) || ROCM_VERSION >= 50300
+#ifdef USE_C10D_NCCL
+  // If the watchdog has remaining work enqueued, an event query on the remaining work will crash
+  // the graph capture.
+  while (!c10d::ProcessGroupNCCL::watchDogsDone()) {
+    TORCH_WARN("Attempting to start graph capture but NCCL ProcessGroup(s) have remaining enqueued work. Waiting for enqueued work to finish...");
+    // Yield
+    std::this_thread::sleep_for(
+        std::chrono::milliseconds(_busy_wait_millis));
+  }
+#endif
   TORCH_CHECK(!has_graph_exec_,
```

Comment:
Don't we have a chance of a data race between these lines, if we start a new NCCL thread ? https://github.com/pytorch/pytorch/blob/36946bc6b5a5b9339c33f51a8446cdd2ba0391e0/aten/src/ATen/cuda/CUDAGraph.cpp#L80-L133

I don't know NCCL, so maybe this is not possible

Reply:
The graph capture begin is intended to be done on the main thread, and if someone is mixing side thread NCCL-usage and graph captures I believe they should be aware of the dangers at that point---there would also be a race also in that there would be no guarantees that their NCCL side thread work would be captured.

I've moved the check to be after the allocator's capture status is updated, just to keep things tidier.